Ex4 chat masking

LCOleksii opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Explanation of Change
<!-- Explain what your change does and how it addresses the linked issue -->
UNMASK:
Chats between Non-Expensify users and Expensify users (e.g. Concierge)

MASK:
Chats between Expensify users and other Expensify users

Chats between non-Expensify users that do NOT include any Expensify users

Exposed PersonalDetailsProvider from OnyxProvider
Added CONST that represents email domain for support team:
- EXPENSIFY_TEAM_EMAIL_DOMAIN

Updated @lib/Fullstory to evaluate MASK/UNMASK based on provided report context

Added isExpensifyAndCustomerChat exported function to ReportUtils
to handle report type evaluation.


### Fixed Issues
<!---
1. Please postfix `$` with a URL link to the GitHub issue this Pull Request is fixing. For example, `$ https://github.com/Expensify/App/issues/<issueID>`.
2. Please postfix  `PROPOSAL:` with a URL link to your GitHub comment, which contains the approved proposal (i.e. the proposal that was approved by Expensify).  For example, `PROPOSAL: https://github.com/Expensify/App/issues/<issueID>#issuecomment-1369752925`

Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the github issue and your comment proposal ; otherwise, the linking and its automation will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<issueID>
$ https://github.com/Expensify/App/issues/<issueID(comment)>

Do NOT only link the issue number like this: $ #<issueID>
--->
$ https://github.com/Expensify/App/issues/52271
PROPOSAL:
Improve MASK/UNMASK process based on report context.
No visual changes. Component attributes added.

### Tests
<!---
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

- [x] Verify that no errors appear in the JS console
- [x] Verify components have appropriate "Fullstory" properties set
QA:
- Navigate using the fullstory@applause email
- Send a message on a Concierge chat
- Send a message on a chat with an expensify user
- create a policy, and add expenses to a policy expense chat
- Check on Fullstory if your session and confirm that you can only see the full data in the policy expense chat

### Offline tests
<!---
Add any relevant steps that validate your changes work as expected in a variety of network states e.g. "offline", "spotty connection", "slow internet", etc. Manual test steps should be written so that your reviewer and QA testers can repeat and verify one or more expected outcomes. If you are unsure how the behavior should work ask for advice in the `#expensify-open-source` Slack channel.
--->

### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

- [x] Verify that no errors appear in the JS console

### PR Author Checklist
<!--
This is a checklist for PR authors. Please make sure to complete all tasks and check them off once you do, or else your PR will not be merged!
-->

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Android: mWeb Chrome</summary>

<!-- add screenshots or videos here -->
$https://drive.google.com/file/d/1KoI9jJIAkUD80-na2Xzf35iFoI1wuLER/view?usp=drive_link

</details>

<details>
<summary>iOS: Native</summary>

<!-- add screenshots or videos here -->
$https://drive.google.com/file/d/10QZ6PG__dkNqF5RbQVFbSbhc4yzbi8Dv/view?usp=drive_link

</details>

<details>
<summary>iOS: mWeb Safari</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

<!-- add screenshots or videos here -->
$https://drive.google.com/file/d/1ZP5xIH7hcxV_LS_iA2PpbZDUgNRVaNjg/view?usp=sharing
</details>

<details>
<summary>MacOS: Desktop</summary>

<!-- add screenshots or videos here -->

</details>

LCOleksii committed:
#EX-4
UNMASK:
Chats between Non-Expensify users and Expensify users (e.g. Concierge)

MASK:
Chats between Expensify users and other Expensify users

Chats between non-Expensify users that do NOT include any Expensify users

Exposed PersonalDetailsProvider from OnyxProvider
Added CONST that represents email domain for support team:
- EXPENSIFY_TEAM_EMAIL_DOMAIN

Updated @lib/Fullstory to evaluate MASK/UNMASK based on provided report context

Added isExpensifyAndCustomerChat exported function to ReportUtils
to handle report type evaluation.

LCOleksii committed:
#EX-4
Clean perfcheck passed.

LCOleksii committed:
#EX-4
Removed unused imports

srikarparsi commented:
Hi @LCOleksii, is this for a specific issue? Please let me know and I can reopen this PR

danieldoglas commented:
hm @srikarparsi why did you close this PR?

srikarparsi commented:
Oh sorry, I didn't know if it was a legit one because I didn't see a linked issue and there was a bunch of lint errors when it was ready for review. But re-opening since you seem to suggest it's legit?

danieldoglas commented:
Yep, it is legit.

srikarparsi commented:
Re-assigning you @danieldoglas if that's okay

LCOleksii committed:
#EX-4
Refactoring, context evaluation moved to component level.
Changed function args usage in related functions.

LCOleksii committed:
#EX-4
ESLINT fixes

LCOleksii committed:
#EX-4
ESLINT fixes

LCOleksii committed:
#EX-4
ESLINT fixes

Prefer an early return to a conditionally-wrapped function body  rulesdir/prefer-early-return

LCOleksii committed:
#EX-4
Prettier fixes

LCOleksii committed:
#EX-4
Prettier fixes

LCOleksii committed:
#EX-4
ESlint double neg check fixes

gedu commented on src/libs/Fullstory/index.ts:
```diff
@@ -91,5 +101,86 @@ const FS = {
+    return fsAttrValue;
+}
+
+function getChatFSAttributes(context: OnyxEntry<PersonalDetailsList>, name: string, report: OnyxInputOrEntry<Report>, prefix: boolean): string {
+    let componentPrefix = prefix ? `${name},` : '';
+    let componentSuffix = name ? `,fs-${name}` : '';
+    let fsAttrValue = '';
+
+    if (isConciergeChatReport(report)) {
+        componentPrefix = prefix ? `${CONCIERGE}-${name},` : '';
+        componentSuffix = name ? `,fs-${name}` : '';
+        /*
+        concierge-chatMessage,fs-unmask,fs-chatMessage
+        */
+        fsAttrValue = `${componentPrefix}${UNMASK}${componentSuffix}`;
+    } else if (isExpensifyAndCustomerChat(context, report)) {
+        componentPrefix = prefix ? `${CUSTOMER}-${name},` : '';
+        componentSuffix = name ? `,fs-${name}` : '';
+        /*
+        customer-chatMessage,fs-unmask,fs-chatMessage
+        */
+        fsAttrValue = `${componentPrefix}${UNMASK}${componentSuffix}`;
+    } else {
+        componentPrefix = prefix ? `${OTHER}-${name},` : '';
+        componentSuffix = name ? `,fs-${name}` : '';
+        /*
+        other-chatMessage,fs-mask,fs-chatMessage
+        */
+        fsAttrValue = `${componentPrefix}${MASK}${componentSuffix}`;
+    }
+
+    return fsAttrValue;
+}
```
I think this logic can be simplified as 
```suggestion
function getChatFSAttributes(context: OnyxEntry<PersonalDetailsList>, name: string, report: OnyxInputOrEntry<Report>, prefix: boolean): string {
    const baseName = name ? `,fs-${name}` : '';
    let componentPrefix = '';
    let maskType = MASK;

    if (isConciergeChatReport(report)) {
        componentPrefix = CONCIERGE;
        maskType = UNMASK;
    } else if (isExpensifyAndCustomerChat(context, report)) {
        componentPrefix = CUSTOMER;
        maskType = UNMASK;
    } else {
        componentPrefix = OTHER;
    }

    const prefixPart = prefix ? `${componentPrefix}-${name},` : '';
    return `${prefixPart}${maskType}${baseName}`;
}
```

gedu commented on src/libs/Fullstory/index.ts:
```diff
@@ -91,5 +101,86 @@ const FS = {
+    return fsAttrValue;
+}
+
+function getChatFSAttributes(context: OnyxEntry<PersonalDetailsList>, name: string, report: OnyxInputOrEntry<Report>, prefix: boolean): string {
```
Would be great to add test on all this functions

gedu commented on src/libs/Fullstory/index.ts:
```diff
@@ -91,5 +101,86 @@ const FS = {
+ * Extract values from non-scraped at build time attribute WEB_PROP_ATTR,
+ * reevaluate "fs-class".
+ */
+function parseFSAttributes(): void {
```
I’m concerned about this function; it seems to be executed multiple times in a row. Each execution tests approximately 300 elements, and this occurs when switching reports, even if switching between the same reports.

<img width="237" alt="Screenshot 2024-11-06 at 19 11 58" src="https://github.com/user-attachments/assets/79a035e2-9a88-400e-8bde-6d54275b3a72">

Would be great to avoid it, given that I felt the report switching lagging

gedu replied:
I also suggest to add test on this function, which seems complex and any future update should be cover by tests

LCOleksii committed:
#EX-4
Performance review fixes,
Cleanup.

gedu commented on src/libs/Fullstory/index.native.ts:
```diff
@@ -63,5 +70,42 @@ const FS = {
+    return `${name},${mask ? MASK : UNMASK}`;
+}
+
+function getChatFSAttributes(context: OnyxEntry<PersonalDetailsList>, name: string, report: OnyxInputOrEntry<Report>): string {
```
did some benchmark and it is better to return earlier, also I think is cleaner
<img width="954" alt="Screenshot 2024-11-07 at 21 14 14" src="https://github.com/user-attachments/assets/110d4c15-9230-4dd6-bcdf-adafbfc8ba4b">

```suggestion
function getChatFSAttributes(context: OnyxEntry<PersonalDetailsList>, name: string, report: OnyxInputOrEntry<Report>): string[] {
    if (!name) {
        return '';
    }

    if (isConciergeChatReport(report)) {
        const formattedName = `${CONCIERGE}-${name}`;
        return [`${formattedName}`, `${UNMASK},${formattedName}`];
    } 

    if (isExpensifyAndCustomerChat(context, report)) {
        const formattedName = `${CUSTOMER}-${name}`;
        return [`${formattedName}`, `${UNMASK},${formattedName}`];
    }

    const formattedName = `${OTHER}-${name}`;
    return [`${formattedName}`, `${MASK},${formattedName}`];
}
```

gedu commented on src/libs/Fullstory/index.ts:
```diff
@@ -92,5 +102,70 @@ const FS = {
+    return `${name}`;
+}
+
+function getChatFSAttributes(context: OnyxEntry<PersonalDetailsList>, name: string, report: OnyxInputOrEntry<Report>): string {
```
Same as my previous suggestion

LCOleksii committed:
#EX-4
Performance review fixes

LCOleksii committed:
#EX-4
TScheck, Prettier, lint fixes

danieldoglas commented on src/libs/ReportUtils.ts:
```diff
@@ -8407,6 +8407,52 @@ function isExpenseReportWithoutParentAccess(report: OnyxEntry<Report>) {
     return isExpenseReport(report) && report?.hasParentAccess === false;
 }
 
+function isExpensifyAndCustomerChat(participantsContext: OnyxEntry<PersonalDetailsList>, report: OnyxInputOrEntry<Report>): boolean {
```
```suggestion
function isExpensifyAndCustomerChat(participantsContext: OnyxEntry<PersonalDetailsList>, report: OnyxInputOrEntry<Report>): boolean {
```

I don't think this is a correct name for this function. We're also checking if the chat is policyExpenseChat and returning true, and in those cases, it's not a `expensify and customer` chat. We should probably rename this to something like `shouldUnmaskChat` or something like that.

LCOleksii replied:
Thank you.
Done changed @danieldoglas

LCOleksii committed:
#EX-4
Better function naming.

LCOleksii committed:
#EX-4
Higher priority to unmask policy expense chats.

LCOleksii commented on src/libs/ReportUtils.ts:
```diff
@@ -8407,6 +8407,52 @@ function isExpenseReportWithoutParentAccess(report: OnyxEntry<Report>) {
+            const id = Number(participantAccountID);
+            const contextAccountData = participantsContext[id];
+            if (contextAccountData) {
+                const login = contextAccountData.login ?? '';
+                if (login.endsWith(CONST.EMAIL.EXPENSIFY_EMAIL_DOMAIN)) {
+                    return true;
+                }
+
+                if (login.endsWith(CONST.EMAIL.EXPENSIFY_TEAM_EMAIL_DOMAIN)) {
+                    return true;
+                }
```
Ok, got ot.

LCOleksii commented on src/libs/ReportUtils.ts:
```diff
@@ -8407,6 +8407,52 @@ function isExpenseReportWithoutParentAccess(report: OnyxEntry<Report>) {
+        }
+    }
+
+    // System users communication
+    const expensifyTeam = new Set(Object.values(CONST.ACCOUNT_ID));
+    const hasIntersection = [...participantAccountIDs].some((id) => expensifyTeam.has(Number(id)));
+    if (hasIntersection) {
+        return true;
+    }
```
Set.Intersection is not available in the TS version you are using.
this is alternative to intersection check.

LCOleksii replied:
I have removed this bit as this was not part of original ticket.

LCOleksii committed:
#EX-4
Internal 1:1 team chat masked.
System users check masked.

gedu commented on src/libs/ReportUtils.ts:
```diff
@@ -8407,6 +8407,51 @@ function isExpenseReportWithoutParentAccess(report: OnyxEntry<Report>) {
+            if (contextAccountData) {
+                const login = contextAccountData.login ?? '';
+
+                if (login.endsWith(CONST.EMAIL.EXPENSIFY_EMAIL_DOMAIN) ?? login.endsWith(CONST.EMAIL.EXPENSIFY_TEAM_EMAIL_DOMAIN)) {
```
What it means here the use of `??`? shouldn't be the OR operator? `||`?

gedu commented on src/libs/Fullstory/index.ts:
```diff
@@ -92,5 +102,73 @@ const FS = {
+    }
+    if (isConciergeChatReport(report)) {
+        const formattedName = `${CONCIERGE}-${name}`;
+        return [`${formattedName},${UNMASK}`, `${formattedName}`];
+    }
+    if (shouldUnmaskChat(context, report)) {
+        const formattedName = `${CUSTOMER}-${name}`;
+        return [`${formattedName},${UNMASK}`, `${formattedName}`];
+    }
+
+    const formattedName = `${OTHER}-${name}`;
+    return [`${formattedName},${MASK}`, `${formattedName}`];
```
there is no need to interpolate `formattedName`, it is an string already, so the second position can be just 
`[`${formattedName}, ${XXXXX}`, formattedName]`

gedu commented on src/pages/home/report/ReportActionsList.tsx:
```diff
@@ -709,13 +712,19 @@ function ReportActionsList({
     // When performing comment linking, initially 25 items are added to the list. Subsequent fetches add 15 items from the cache or 50 items from the server.
     // This is to ensure that the user is able to see the 'scroll to newer comments' button when they do comment linking and have not reached the end of the list yet.
     const canScrollToNewerComments = !isLoadingInitialReportActions && !hasNewestReportAction && sortedReportActions.length > 25 && !isLastPendingActionIsDelete;
+    const [reportActionsListTestID, reportActionsListFSClass] = getChatFSAttributes(participantsContext, 'ReportActionsList', report);
+
```
`ReportActionsList` is being rerender many times, when switching and when sending/receiving  messages, I would recommend to wrap it in a `useMemo` and the dependency array would be `[report.reportID]`
<img width="292" alt="Screenshot 2024-11-13 at 11 44 20" src="https://github.com/user-attachments/assets/0da8b010-aa07-444c-8084-c716492e1c9f">

<img width="304" alt="Screenshot 2024-11-13 at 11 45 56" src="https://github.com/user-attachments/assets/92e7e769-791f-4a12-8ea8-d6099dddc454">

LCOleksii committed:
#EX-4
clean unnecessary type coercion

danieldoglas commented:
@LCOleksii bump for the feedbacks above.

danieldoglas commented:
@LCOleksii please merge main into this

LCOleksii committed:
Merge remote-tracking branch 'upstream/main' into ex4-chat-masking

# Conflicts:
#	src/libs/ReportUtils.ts
#	src/pages/home/report/ReportActionsList.tsx

LCOleksii committed:
#EX-4
Error fix: Prettier diff detected

gedu commented:
There is a failing test

gedu commented:
Any news here?

danieldoglas commented:
@LCOleksii bump on te test fix

LCOleksii committed:
#EX-4
Error fix: Perf-test missing mocked library fix.

LCOleksii committed:
#EX-4
Eslint check fix.

danieldoglas commented:
@gedu ready for your eyes again

gedu commented:
Cool, I will take a look tomorrow

danielrvidal commented:
@gedu this is still being worked on, correct? Who are you waiting on?

gedu commented:
Sorry, something with more priority shows up, tomorrow first thing in the morning I will check it

gedu commented:
LGTM

danieldoglas committed:
Merge branch 'main' into ex4-chat-masking
